src/Raven.Server/Documents/Sharding/Executors/AbstractExecutor.cs

```diff
@@ -95,6 +95,21 @@ public Task<TResult> ExecuteParallelForAllAsync<TResult>(IShardedOperation<TResu
             {
                 try
                 {
+                    if (command.CommandTask.IsCompleted == false)
+                    {
+                        // we must not return the context if a command task is still running so we need to await it
+                        // this can happen when ThrowOnFailure mode is used and we throw on first failure
+
+                        try
+                        {
+                            await command.CommandTask;
```

Comment:
maybe it would be better to dispose as a continuation? so we will not wait here

Reply:
I'm not sure if I want to delegate the return of the context to the background. I prefer to have it released explicitly. It's not going to be a common scenario.

Reply:
Another option - maybe instead of waiting we should cancel those requests / tasks first?

Reply:
Merged, will create an issue to address my comment.. some time in the future :)

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-20418